### ISSUE TITLE
"C-x k RET" leads me to a crash

### ISSUE THREAD

**Issue by YoungFrog** [atlas-engineer/nyxt#1546]
It was my first time launching nyxt, I played with it for a few minutes, and then it crashed.
I tried to reproduce the crash, ~without success~ and managed to do it (see below the stack trace) 

I was visiting a results page of google and tried to kill the buffer using "C-x k RET" (I blame muscle memory). During the key combo, nyxt vanished.  Here's the full log from the console after my initial crash (a shorter one follows) :

## Full log and stacktrace

```
$ nyxt --version
Nyxt version 2.1.0

$ nyxt 
<INFO> [06:48:13] Listening to socket "/run/user/1000/nyxt/nyxt.socket".
Nyxt version 2.1.0
<INFO> [06:48:31] Update slot DEFAULT-MODES to (APPEND '(NYXT::EMACS-MODE) %SLOT-DEFAULT%).
<WARN> [06:48:31] Warning: Error on separate thread: The file
                          #P"/home/youngfrog/.config/nyxt/auto-config.lisp"
                          does not exist:
                            Aucun fichier ou dossier de ce type
<INFO> [06:49:17] Update slot DEFAULT-MODES to (APPEND '(NYXT::EMACS-MODE) %SLOT-DEFAULT%).
<INFO> [06:49:17] Appending to "/home/youngfrog/.config/nyxt/auto-config.lisp":
(DEFINE-CONFIGURATION BUFFER
  ((DEFAULT-MODES (APPEND '(NYXT::EMACS-MODE) %SLOT-DEFAULT%))))
<WARN> [06:49:40] Warning: Error on separate thread: There is no applicable method for the generic function
                            #<STANDARD-GENERIC-FUNCTION HISTORY-TREE:PARENT (1)>
                          when called with arguments
                            (NIL).
See also:
  The ANSI Standard, Section 7.6.6
<WARN> [06:49:45] Warning: Error on separate thread: There is no applicable method for the generic function
                            #<STANDARD-GENERIC-FUNCTION HISTORY-TREE:PARENT (1)>
                          when called with arguments
                            (NIL).
See also:
  The ANSI Standard, Section 7.6.6
<INFO> [06:50:19] Pressed keys: C-s
<INFO> [06:50:23] Pressed keys: C-h
<INFO> [06:50:31] Pressed keys: C-h
<INFO> [06:50:34] Press a key sequence to describe (cancel with 'escape escape'):
<INFO> [06:50:39] Press a key sequence to describe (cancel with 'escape escape'): "C-s" [C-#39]
<INFO> [06:50:55] Pressed keys: C-x
<INFO> [06:50:58] Pressed keys: C-x
<INFO> [06:50:59] Pressed keys: C-x
<INFO> [06:51:00] Pressed keys: C-x
<INFO> [06:51:44] Pressed keys: C-x
<INFO> [06:53:34] Loading "<redacted>".
<INFO> [06:53:35] Finished loading "<redacted>".
<INFO> [06:53:38] Pressed keys: C-x
<INFO> [06:53:50] Pressed keys: C-h
<INFO> [06:53:52] Pressed keys: C-s
<INFO> [06:53:54] Pressed keys: C-s
<INFO> [06:54:20] Pressed keys: C-x
<INFO> [06:54:21] Pressed keys: C-x
<INFO> [06:54:25] Pressed keys: M-g
<INFO> [06:54:37] Loading "<redacted>".
<INFO> [06:54:38] Finished loading "<redacted>".
<INFO> [06:54:48] Loading "<redacted>".
<INFO> [06:54:49] Finished loading "<redacted>".
<INFO> [06:54:55] Pressed keys: C-x
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING
                                    {1008B4A503}>:
  Owner with identifier "" does not exist

Backtrace for: #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING {1008B4A503}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SIMPLE-ERROR "Owner with identifier ~s does not exist" {1008B8F5F3}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK SB-EXT:*INVOKE-DEBUGGER-HOOK* #<SIMPLE-ERROR "Owner with identifier ~s does not exist" {1008B8F5F3}>)
2: (INVOKE-DEBUGGER #<SIMPLE-ERROR "Owner with identifier ~s does not exist" {1008B8F5F3}>)
3: (ERROR "Owner with identifier ~s does not exist" "")
4: (HISTORY-TREE:SET-CURRENT-OWNER #<HISTORY-TREE:HISTORY-TREE {100657D713}> "")
5: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN WINDOW-SET-BUFFER))
6: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
7: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<FUNCTION (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN WINDOW-SET-BUFFER) {7F5D0CDAE7AB}> #<SB-THREAD:MUTEX "Anonymous recursive lock" owner: #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING {1008B4A503}>> T NIL)
8: (WINDOW-SET-BUFFER #<USER-WINDOW {10065774F3}> #<USER-WEB-BUFFER {100A12AA23}> :FOCUS NIL)
9: (SET-CURRENT-BUFFER #<USER-WEB-BUFFER {100A12AA23}> :FOCUS NIL)
10: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
11: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
12: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::RUN))
13: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
14: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::RUN))
15: (SB-THREAD::RUN)
16: ("foreign function: call_into_lisp")
17: ("foreign function: funcall1")

unhandled condition in --disable-debugger mode, quitting
<WARN> [06:54:55] Warning: Error: Joining thread failed: thread #<THREAD "Nyxt renderer thread" ABORTED {100807AA53}> did not return normally.
```

## Reproducing

I start from scratch using ```rm -rf ~/.local/share/nyxt ~/.config/nyxt```

I then create auto-config.lisp file, which will allow me to click on "Settings" then "Use emacs"
(otherwise a WARN is thrown -- see the first WARN from the log above), and run nyxt:

```
mkdir ~/.config/nyxt; touch ~/.config/nyxt/auto-config.lisp; nyxt
```
Inside Nyxt, do the following:
click "Common Settings", then "Use Emacs"
hit `M-l` and type `google.com` then `RET`
once the page has loaded, hit in rapid succession: `C-x k RET`
nyxt crashes for me.

### Backtrace for the repro

```
$ rm -rf /home/youngfrog/.local/share/nyxt ~/.config/nyxt; mkdir -p ~/.config/nyxt; touch ~/.config/nyxt/auto-config.lisp; nyxt
<INFO> [09:41:15] Listening to socket "/run/user/1000/nyxt/nyxt.socket".
Nyxt version 2.1.0
<INFO> [09:41:15] Loading Lisp file "/home/youngfrog/.config/nyxt/auto-config.lisp".
<INFO> [09:41:18] Update slot DEFAULT-MODES to (APPEND '(NYXT::EMACS-MODE) %SLOT-DEFAULT%).
<INFO> [09:41:18] Appending to "/home/youngfrog/.config/nyxt/auto-config.lisp":
(DEFINE-CONFIGURATION BUFFER
  ((DEFAULT-MODES (APPEND '(NYXT::EMACS-MODE) %SLOT-DEFAULT%))))
<INFO> [09:41:22] Loading "https://google.com/".
<INFO> [09:41:22] Finished loading "https://www.google.com/#spf=1624434082342".
<INFO> [09:41:24] Pressed keys: C-x
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING
                                    {1007336D83}>:
  Owner with identifier "" does not exist

Backtrace for: #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING {1007336D83}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SIMPLE-ERROR "Owner with identifier ~s does not exist" {10072AD4C3}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK SB-EXT:*INVOKE-DEBUGGER-HOOK* #<SIMPLE-ERROR "Owner with identifier ~s does not exist" {10072AD4C3}>)
2: (INVOKE-DEBUGGER #<SIMPLE-ERROR "Owner with identifier ~s does not exist" {10072AD4C3}>)
3: (ERROR "Owner with identifier ~s does not exist" "")
4: (HISTORY-TREE:SET-CURRENT-OWNER #<HISTORY-TREE:HISTORY-TREE {1009F13B83}> "")
5: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN WINDOW-SET-BUFFER))
6: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
7: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<FUNCTION (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN WINDOW-SET-BUFFER) {7F85285A67AB}> #<SB-THREAD:MUTEX "Anonymous recursive lock" owner: #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING {1007336D83}>> T NIL)
8: (WINDOW-SET-BUFFER #<USER-WINDOW {1008BB30D3}> #<USER-WEB-BUFFER {1008DEBD43}> :FOCUS NIL)
9: (SET-CURRENT-BUFFER #<USER-WEB-BUFFER {1008DEBD43}> :FOCUS NIL)
10: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
11: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
12: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::RUN))
13: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
14: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::RUN))
15: (SB-THREAD::RUN)
16: ("foreign function: call_into_lisp")
17: ("foreign function: funcall1")

unhandled condition in --disable-debugger mode, quitting
<WARN> [09:41:24] Warning: Error: Joining thread failed: thread #<THREAD "Nyxt renderer thread" ABORTED {10082054D3}> did not return normally.

```

## System information

```
$ nyxt --system-information
Nyxt version: 2.1.0
Renderer version: GI-GTK
Operating system kernel: Linux 5.9.0-050900-generic
Lisp implementation: SBCL 2.1.0.0.HEAD.1-46cb065
Features: (WEBKIT2 WEBKIT2-2.32 WEBKIT2-PASTE-PLAINTEXT WEBKIT2-TRACKING WEBKIT2-MUTE
 WEBKIT2-EMOJI WEBKIT2-MEDIA WEBKIT2-SANDBOXING GTK-3-22 GTK-3-20 GTK-3-18
 GTK-3-16 GTK-3-14 GTK-3-12 GTK-3-10 GTK-3-8 GTK-3-6 GTK-3-4 GTK GDK-3-22
 GDK-3-20 GDK-3-18 GDK-3-16 GDK-3-14 GDK-3-12 GDK-3-10 GDK-3-8 GDK-3-6 GDK-3-4
 CAIRO-1-10 CAIRO-1-12 GDK-PIXBUF GLIB-2-30 GLIB-2-32 GLIB-2-34 GLIB-2-36
 GLIB-2-38 GLIB-2-40 GLIB-2-42 GLIB-2-44 GLIB-2-46 GLIB-2-48 GLIB-2-50
 GLIB-2-52 GLIB-2-54 GLIB-2-56 GLIB-2-58 GLIB NYXT-2 FSET-EXT-STRINGS
 CUSTOM-HASH-TABLE-NATIVE SWANK PLUMP-UTF-32 GLOBAL-VARS DECLARE-TYPES
 PARENSCRIPT NAMED-READTABLES 21BIT-CHARS CHUNGA FLEXI-STREAMS CLOSER-MOP
 CL-PPCRE-UNICODE CL-UNICODE BORDEAUX-THREADS SPLIT-SEQUENCE GRAY-STREAMS
 FAST-IO-SV FAST-IO SBCL-USES-SB-ROTATE-BYTE ASDF-SYSTEM-CONNECTIONS
 CL-JSON-CLOS CL-JSON THREAD-SUPPORT CL-PPCRE FLAT-NAMESPACE X86-64 UNIX CFFI
 FLAT-NAMESPACE SEQUENCE-EMPTYP QUICKLISP ASDF3.3 ASDF3.2 ASDF3.1 ASDF3 ASDF2
 ASDF OS-UNIX NON-BASE-CHARS-EXIST-P ASDF-UNICODE X86-64 GENCGC 64-BIT ANSI-CL
 COMMON-LISP ELF IEEE-FLOATING-POINT LINUX LITTLE-ENDIAN
 PACKAGE-LOCAL-NICKNAMES SB-CORE-COMPRESSION SB-LDB SB-PACKAGE-LOCKS SB-THREAD
 SB-UNICODE SBCL UNIX)

ASDF version: 3.3.4
Critical dependencies: (/home/runner/work/nyxt/nyxt/_build/submodules/cl-cffi-gtk/gtk/cl-cffi-gtk.asd
 /home/runner/work/nyxt/nyxt/_build/quicklisp-client/dists/quicklisp/software/cl-gobject-introspection-20210124-git/cl-gobject-introspection.asd
 /home/runner/work/nyxt/nyxt/_build/submodules/cl-webkit/webkit2/cl-webkit2.asd)

Quicklisp dist version: 2021-05-31
Quicklisp client version: 2020-01-04
Local project directories: (/home/runner/work/nyxt/nyxt/_build/quicklisp-client/local-projects/)
Critical dependencies(#<SYSTEM cl-cffi-gtk / cl-cffi-gtk-20201220-git / quicklisp 2021-05-31>
 #<SYSTEM cl-gobject-introspection / cl-gobject-introspection-20210124-git / quicklisp 2021-05-31>
 #<SYSTEM cl-webkit2 / cl-webkit-20210531-git / quicklisp 2021-05-31>)
```

**Comment by Ambrevar** [comment]
Thanks for reporting!

I've fixed the issue with non-existing files in
4dd621a943260d94e0c3caeb946662e910989e0d.

I can't reproduce your issue with buffer deletion, but it looks like a
race-condition with how we delete buffer and focus on the next one.

Looking at the backtrace:

> 3: (ERROR "Owner with identifier ~s does not exist" "")
> 4: (HISTORY-TREE:SET-CURRENT-OWNER #<HISTORY-TREE:HISTORY-TREE {100657D713}> "")
> 5: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN WINDOW-SET-BUFFER))
> 6: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
> 7: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<FUNCTION (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN WINDOW-SET-BUFFER) {7F5D0CDAE7AB}> #<SB-THREAD:MUTEX "Anonymous recursive lock" owner: #<SB-THREAD:THREAD "Prompter follow selection thread" RUNNING {1008B4A503}>> T NIL)
> 8: (WINDOW-SET-BUFFER #<USER-WINDOW {10065774F3}> #<USER-WEB-BUFFER {100A12AA23}> :FOCUS NIL)
> 9: (SET-CURRENT-BUFFER #<USER-WEB-BUFFER {100A12AA23}> :FOCUS NIL)

We see that (id buffer) is "" for `#<USER-WEB-BUFFER {100A12AA23}>`.

The id is "" because of `add-to-recent-buffers`.

So if I get it right, 2 `buffer-delete` are run in parallel when you
delete buffers fast enough, and the race condition happens at
`get-inative-buffers`.

I'll work on it.

**Comment by Ambrevar** [comment]
After some investigation, this is not very obvious to solve.  We could add a lock in `buffers-delete`, but that would be just a band aid for a very large wound.

Any time we access and manipulate buffers in parallel, we run at the risk of manipulating a buffer that gets deleted in another thread.

How can we prevent this?  Each time we deal with a buffer, would we have to lock it?  Does not seem easy.

The main victim of these race conditions is `htree:set-current-owner`, so as a workaround we could at least lock this.

Any idea?  @jmercouris @aartaka

**Comment by aartaka** [comment]
Maybe filter buffers by being dead in the calling code in the meantime?

Other than that, I'm out of ideas :(

**Comment by jmercouris** [comment]
Perhaps we should not have a concept of 'dead' buffers, that would solve this issue.

**Comment by jmercouris** [comment]
Then we would rely only on the Lisp GC to actually remove the buffer, and then "deleting" would just remove a buffer from the buffers list.

**Comment by Ambrevar** [comment]
@jmercouris: You may have misunderstood the issue, it's not related to
dead buffers or garbage collection.

I'll try to clarify:

A buffer may get _semantically_ deleted (implementation does not matter
here) in one thread while being
manipulated in another.

Manipulating a deleted buffer poses an integrity issue (e.g. with
history, buffer list, window display, etc.).

How do we ensure this won't blow up?

Example:
- Thread A deletes buffer 17.
- Thread B commands current window to display buffer 17.

Possible result: window displays buffer 17, which is actually a deleted
buffer (and because of WebKit, it may display just a white canvas).

**Comment by Ambrevar** [comment]
Artyom Bologov ***@***.***> writes:

> Maybe filter buffers by being dead in the calling code in the meantime?

Still prone to a race condition.  Unless we put a lock on the buffer
list.

We start putting locks everywhere, but that does not seem right.

**Comment by Ambrevar** [comment]
Maybe it's not that hard after all, and we just need to make the
inventory of where buffers happen.

- In the buffer list `(buffers *browser*)`.
- As the window's current buffer being displayed (`active-buffer` slot).

Anything else?

We can also manipulate buffers more functionally, for instance never
remove their ID among other things.  Is there such a thing as immutable
objects in CL?
Otherwise, we could add a lock on all setf methods for the buffer.  Is
there a convenient way to do this?

With that and locks on the buffer list, etc., it should make buffer
access and manipulation much more reliable.

Thoughts?

**Comment by Ambrevar** [comment]
Maybe cl-gserver can help here: https://github.com/mdbergmann/cl-gserver/issues/33

**Comment by aartaka** [comment]
@Ambrevar, I am able to reproduce the buffer kill hang reliably with 15+ buffers killed at the same time.

As a side note, some fun anomalies of power-user browsing:
- Github logs you out after the fifth buffer opened too fast for a human xD
- Github issues are perfect for testing the multiple buffer kill/creation, as you can simply type `/issues/` in the `follow-hint-new-buffer` prompt and it will list all the issue links on the page :)

**Comment by Ambrevar** [comment]
For now, we can add a global lock to delete-buffer with a sleep,
say 0.1s, between two calls.

Maybe we can do the same with follow-hint.

For 3.0, I propose we introduce actors to solve this problem centrally
(and more generally).

**Comment by aartaka** [comment]
> For now, we can add a global lock to delete-buffer with a sleep, say 0.1s, between two calls. Maybe we can do the same with follow-hint. For 3.0, I propose we introduce actors to solve this problem centrally (and more generally).

Yes, sounds sane.

**Comment by jmercouris** [comment]
You are correct, I had misunderstood the issue. I see now. Pierre and I also discussed utilizing a queue in the event that part of the reason for crashes is due to WebKit constraints.

**Comment by Ambrevar** [comment]
I can't reproduce neither on master nor on `2-series`.
@aartaka Can you still reproduce?
Can you share with me an exact recipe with which URLs you've used for testing?

**Comment by aartaka** [comment]
> I can't reproduce neither on master nor on `2-series`.
> @aartaka Can you still reproduce?
> Can you share with me an exact recipe with which URLs you've used for testing?

Sure! The list that Nyxt reliably hangs trying to close:
``` lisp
'("https://github.com/atlas-engineer/nyxt/issues/1938"
 "https://github.com/atlas-engineer/nyxt/issues/1937"
 "https://github.com/atlas-engineer/nyxt/issues/1935"
 "https://github.com/atlas-engineer/nyxt/issues/1923"
 "https://github.com/atlas-engineer/nyxt/issues/1918"
 "https://github.com/atlas-engineer/nyxt/issues/1915"
 "https://github.com/atlas-engineer/nyxt/issues/1914"
 "https://github.com/atlas-engineer/nyxt/issues/1913"
 "https://github.com/atlas-engineer/nyxt/issues/1912"
 "https://github.com/atlas-engineer/nyxt/issues/1910"
 "https://github.com/atlas-engineer/nyxt/issues/1909"
 "https://github.com/atlas-engineer/nyxt/issues/1908"
 "https://github.com/atlas-engineer/nyxt/issues/1907"
 "https://github.com/atlas-engineer/nyxt/issues/1906"
 "https://github.com/atlas-engineer/nyxt/issues/1905"
 "https://github.com/atlas-engineer/nyxt/issues/1903"
 "https://github.com/atlas-engineer/nyxt/issues/1901"
 "https://github.com/atlas-engineer/nyxt/issues/1896"
 "https://github.com/atlas-engineer/nyxt/issues/1895"
 "https://github.com/atlas-engineer/nyxt/issues/1893"
 "https://github.com/atlas-engineer/nyxt/issues/1892"
 "https://github.com/atlas-engineer/nyxt/issues/1891"
 "https://github.com/atlas-engineer/nyxt/issues/1887"
 "https://github.com/atlas-engineer/nyxt/issues/1886"
 "https://github.com/atlas-engineer/nyxt/issues/1883"
 "https://github.com/atlas-engineer/nyxt/issues/1908"
 "https://github.com/atlas-engineer/nyxt/issues/1906"
 "https://github.com/atlas-engineer/nyxt/issues/1883"
 "https://github.com/atlas-engineer/nyxt/issues/1901"
 "https://github.com/atlas-engineer/nyxt/issues/1895"
 "https://github.com/atlas-engineer/nyxt/issues")
```

**Comment by aartaka** [comment]
The recipe is simple: open 10+ (or 15+, or 20+, the number I'm not sure about) buffers and try to delete all of them, marking all of them in `delete-buffer` prompt.

The more detailed recipe for exactly how I just tested it:
- Start Nyxt.
- Navigate to `https://github.com/atlas-engineer/nyxt/issues`
- Run `follow-hint-new-buffer-focus` and select all the issue links (list above).
- Hit Return and let those load.
  - I'm not sure that all of those did load, so that half-loaded state can be a breakage reason. I'm pretty sure it isn't, though, as I did reproduce the hang with all the buffers fully loaded before.
- Execute `delete-buffer` (either from `execute-command` prompt or with `C-x k` in Emacs/VI bindings).
- Choose all the buffers except for the `https://github.com/atlas-engineer/nyxt/issues` one.
- Hit Return.
- Observe Nyxt hanging.

**Comment by Ambrevar** [comment]
Thanks!

First finding: I can reproduce with you list of 31 buffers.
With my list of 32 buffers, most of which are on my website, I cannot.
So it seems that the buffer content matters.

My supposition is that WebKitGTK takes more time to delete a heavy
buffer which is what triggers the race condition.

**Comment by Ambrevar** [comment]
There was actually a race condition in history access (due to how the data-storage library works, high time we fixed it).  I've fixed it in #1940 I believe.

**Comment by Ambrevar** [comment]
Fixed is in 2.2.4!
